DOCS-6436 Add Local and Remote Testing Documentation (#20231)
DataDog/documentation: content/en/continuous_testing/environments/proxy_firewall_vpn.md

---

> title: Testing While Using Proxies, Firewalls, or VPNs

Reviewer: ```suggestion
title: Testing with Proxies, Firewalls, or VPNs
```

  tag: "Blog"
  text: "Incorporate Datadog Continuous Testing tests into your CI/CD pipeline"
- link: "https://www.datadoghq.com/blog/internal-application-testing-with-datadog/"
  tag: "Blog"
  text: "Test internal applications with Datadog's testing tunnel and private locations"
- link: "/continuous_testing/cicd_integrations"
  tag: "Documentation"
  text: "Learn about Continuous Testing and CI/CD"
---

## Overview

Most of the development cycle happens within private networks, which are usually inaccessible to Synthetic tests. With the help of [`datadog-ci`][2], you can establish a Continuous Testing tunnel that allow Synthetics Workers to reach the environments that your application is deployed on during the development cycle, such as your development laptop, a CI job, or a private staging environment.

Datadog recommends using the testing tunnel if you need to launch Continuous Testing tests against local versions of your application without deploying a dedicated, long-lasting probing system such as a [private location][1]. You can also use the testing tunnel to trigger tests on short-lived cloud environments.

## What is the testing tunnel?

> The testing tunnel is a functionality that comes with the [@datadog/datadog-ci][2] NPM package which is one of the methods Datadog provides to include your Synthetic tests as part of your CI/CD pipelines.

Reviewer 2: ```suggestion
The testing tunnel is a functionality that comes with the [@datadog/datadog-ci][2] NPM package, which is one of the methods Datadog provides to include your Synthetic tests as part of your CI/CD pipelines.
```

The testing tunnel creates short-lived secure connections between your internal environments and the Datadog infrastructure, allowing you to swiftly trigger Synthetic HTTP and Browser tests on your private applications. This allows Datadog to access and test your internal applications.

{{< img src="continuous_testing/continuous_testing_tunnel.png" alt="Continuous Testing tunnel allows the Synthetics Worker to reach your private applications" width="100%" >}}

First, `datadog-ci` gets a pre-signed URL from Datadog for authentication. Then, it opens a WebSocket secure (WSS) connection to Datadog's managed locations using the pre-signed URL. By using SSH connections through the WebSocket connection, tests are triggered by `datadog-ci` and executed through Datadog's managed locations.

Because DNS resolution is performed through the testing tunnel, you can test applications with internal domains or even on the `localhost` of the machine running `datadog-ci`.

When using the testing tunnel, your tests' locations are overridden by a location that depends on your Datadog account region.

## How to use the testing tunnel

> As mentioned above, the testing tunnel comes with the [`@datadog/datadog-ci`][2] NPM package and is available from version [v0.11.0][3] of the package. To get started, see [Continuous Testing and CI/CD][4].

Reviewer 3: I think the link to that version isn't really useful, and could mislead users into using this version explicitly, which is very outdated.

Also "available from version" could be misinterpreted as "only available from the version `v0.11.0` specifically". That's a bit farfetched, but I think something like `>=`, or "available **starting** from" would be more explicit 😁 

```suggestion
As mentioned above, the testing tunnel comes with the [`@datadog/datadog-ci`][2] NPM package and is available in versions `>=v0.11.0` of the package. To get started, see [Continuous Testing and CI/CD][4].
```
